issue-121: Adds option to have custom location pointers and tooltip for map facet

nikhilj13 opened the pull request:
Adds option to have custom image/Glyphicon as location pointers, also shows a tooltip with the count when you hover over a hot-spot.

For example, below is the map for customers in the United States:
![mapImprovement](https://user-images.githubusercontent.com/31805650/56767420-b6609f00-6779-11e9-91b0-3df4c3c3eeed.gif)

nikhilj13 committed:
Added custom pointers and tooltips for map

luigivaldez commented on src/components/MapFacetContents.js:
```diff
@@ -31,11 +31,20 @@ type MapFacetContentsProps = {
+  /** Custom location pointer glyphicon, default is the map-marker glyphicon */
+  glyph: string | null;
+  /** Custom location pointer image uri, if available image is shown instead of glyphicon */
+  uri: string | null;
```
Can you rename this to "pointerImageUri"? And rename glyph to "pointerGlyph"?

nikhilj13 replied:
Done

luigivaldez commented on src/components/MapFacetContents.js:
```diff
@@ -31,11 +31,20 @@ type MapFacetContentsProps = {
+  /** Custom location pointer image uri, if available image is shown instead of glyphicon */
+  uri: string | null;
+  /** Custom tooltip text to convey what the location pointers are for, for e.g. customers, transaction location etc */
+  tooltip: string;
```
We should let people pass a string that is understood by the `StringUtils.fmt()` method and pass the count as the value parameter. (And then call that method later...)
This would allow the same map's tooltips to say things like "There are 4 stores here" and "There is only one store here" if the count is 4 or 1, respectively.
You could theoretically prepend the number like you're doing now if the string isn't a format string (i.e. if the format string before and after calling `StringUtild.fmt()` is the same...).
In any case, update the comment here to describe it in a bit more detail.

nikhilj13 replied:
That makes sense, I wasn't aware of that function in StringUtils. I changed the tooltips to use StringUtils.fmt() and updated the comment as well.

luigivaldez commented on src/components/MapFacetContents.js:
```diff
@@ -53,6 +62,9 @@ class MapFacetContents extends React.Component<MapFacetContentsDefaultProps, Map
     mapboxKey: '',
+    glyph: 'map-marker',
+    uri: null,
+    tooltip: 'occurrence(s)',
```
We could update this default value to be `'{} occurrence|{} occurrences'` so it would handle the single/plural automatically.

nikhilj13 replied:
Done

luigivaldez commented on src/components/MapFacetContents.js:
```diff
@@ -210,7 +228,9 @@ class MapFacetContents extends React.Component<MapFacetContentsDefaultProps, Map
             style={{ cursor: 'pointer' }}
           >
-            <Glyphicon glyph="map-marker" style={{ fontSize: '18px', color: '#2a689c' }} />
+            <OverlayTrigger overlay={<Tooltip id="tooltip-bottom">{bucket.count} {this.props.tooltip} in this region</Tooltip>}>
```
It would be more readable to put the <Tooltip> tag on a new line, like this:
```
<OverlayTrigger overlay={(
  <Tooltip id="tooltip-bottom">
    {formattedTooltip}
  </Tooltip>
}}> 
  {locationPointer}
</OverlayTrigger>
```

nikhilj13 replied:
Done

luigivaldez commented on src/components/MapFacetContents.js:
```diff
@@ -198,6 +210,12 @@ class MapFacetContents extends React.Component<MapFacetContentsDefaultProps, Map
+      }
       const points = this.props.buckets.map((bucket) => {
         const value = bucket.value; // JSON.parse(bucket.value);
         // Keep track of the boundaries of the coordinates
```
Here is where you might add the tooltip calculation, like:
```
let formattedTooltip;
formattedTooltip = StringUtils.fmt(this.props.tooltip, bucket.count);
if (formattedTooltip === this.props.tooltip) {
  formattedTooltip = `${bucket.count} ${this.props.tooltip}`;    
}
```

nikhilj13 committed:
Completed required changes

- Used StringUtils.fmt for the tooltip
- Changed the names of props

nikhilj13 committed:
Create tree.png

added image used for MapFacetContents'
documentation

luigivaldez commented on src/components/MapFacetContents.js:
```diff
@@ -31,19 +31,25 @@ type MapFacetContentsProps = {
+   * It should specify the singular and plural forms to be used
+   * For example a tooltip for cats would be '{} cat|{} cats'
+   * If the string is not in the format specified above,
+   * it would be used as it is concatenated with the count.
```
I might just change "concatenated with" to be "appended to" so it's unambiguous that the order will be first the count and then the tooltip string.

nikhilj13 replied:
Changed!

luigivaldez reviewed (changes requested):
Last nitpicky thing. :)

nikhilj13 committed:
Update MapFacetContents.js

Changed "concatenated with" to be "appended to" so it's unambiguous that the order will be first the count and then the tooltip string.

klk1010 committed:
Merge branch 'master' into map-improvement

luigivaldez committed:
Tweak comments

luigivaldez committed:
Merge branch 'master' into map-improvement

luigivaldez committed:
Merge branch 'master' into map-improvement

luigivaldez committed:
Fix warning about "possible undefined"
